### PR TITLE
feat(arith): add ConstIntBoundAnalyzer for integer bound propagation

### PR DESCRIPTION
**Pull request by Hzfengsy**
## Summary

Implements constant integer bound propagation through expression trees (PR 2 of arithmetic simplifier series, tracking issue #668).

- **`ConstIntBound`** struct with overflow-safe sentinels (`kPosInf`/`kNegInf = -kPosInf`)
- **`ConstIntBoundAnalyzer`** with `Bind`/`Update`/`EnterConstraint` APIs
- Bound propagation for all expression types via `ExprFunctor<ConstIntBound>` visitor: arithmetic (`Add`/`Sub`/`Mul`/`FloorDiv`/`FloorMod`/`Pow`), comparisons, bitwise, unary (`Neg`/`Abs`/`Cast`), `Min`/`Max`
- Python binding and type stubs for `ConstIntBound`/`ConstIntBoundAnalyzer`
- **46 tests** covering bound analysis for all supported operations

Third-party attribution: takes reference from Apache TVM (`tvm.arith`) and MLC-Python (`mlc.sym`), both Apache License 2.0.

## Test plan

- [x] All 46 new ConstIntBound tests pass
- [x] Full test suite (2765 tests) passes with no regressions
- [x] Build succeeds cleanly
- [x] Pre-commit hooks pass (check-headers, clang-format, cpplint, pyright, ruff)
- [x] Clang-tidy passes with zero errors on changed files

### CODE REVIEW

File: src/ir/arith/const_int_bound.cc [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+  }
+  if (MulWouldOverflow(a, b)) {
+    return pos_result ? Bound::kPosInf : Bound::kNegInf;
+  }
+  return a * b;
+}
+
+static int64_t InfAwareNeg(int64_t a) {
+  if (a == Bound::kPosInf) return Bound::kNegInf;
+  if (a == Bound::kNegInf) return Bound::kPosInf;
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/arith/const_int_bound.cc:75]: Fixed in `a8349e3` — added `NegWouldOverflow` guard that saturates to `kPosInf` when negating `INT64_MIN`. Also added test `test_neg_int64_min`.

File: src/ir/arith/const_int_bound.cc [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+    int64_t e = exp.min_value;
+    if (e == 0) return {1, 1};
+    if (e == 1) return base;
+    if (base.is_non_negative()) {
+      // Monotonically increasing for non-negative base
+      int64_t lo = 1, hi = 1;
+      for (int64_t i = 0; i < e; ++i) {
+        lo = InfAwareMul(lo, base.min_value);
+        hi = InfAwareMul(hi, base.max_value);
+      }
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/arith/const_int_bound.cc]: Fixed in `a8349e3` — replaced O(e) loop with O(log e) exponentiation by squaring via `InfAwarePow` helper. Also added test `test_large_pow_exponent` that verifies fast computation with exponent=1000000.

File: src/ir/arith/const_int_bound.cc [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+    }
+    // Lt: left < right
+    if (auto lt = As<Lt>(expr)) {
+      if (auto var = As<Var>(lt->left_)) {
+        auto rb = VisitExpr(lt->right_);
+        if (rb.is_const()) TryTighten(var.get(), {Bound::kNegInf, rb.max_value - 1});
+      } else if (auto var = As<Var>(lt->right_)) {
+        auto lb = VisitExpr(lt->left_);
+        if (lb.is_const()) TryTighten(var.get(), {lb.min_value + 1, Bound::kPosInf});
+      }
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/arith/const_int_bound.cc:440]: Fixed in `a8349e3` — all ±1 adjustments in `EnterConstraint` now use `InfAwareAdd` instead of raw arithmetic to safely handle sentinel values.

File: src/ir/arith/const_int_bound.cc [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+  Bound VisitExpr_(const CastPtr& op) override {
+    auto a = VisitExpr(op->operand_);
+    // For Cast to integer type, intersect with the target type's representable range
+    auto scalar_type = std::dynamic_pointer_cast<const ScalarType>(op->GetType());
+    if (!scalar_type || !scalar_type->dtype_.IsInt()) return Everything();
+    size_t bits = scalar_type->dtype_.GetBit();
+    if (bits == 0 || bits > 63) return a;  // Unknown or >= 64 bits: pass through
+    int64_t type_min = -(static_cast<int64_t>(1) << (bits - 1));
+    int64_t type_max = (static_cast<int64_t>(1) << (bits - 1)) - 1;
+    return {std::max(a.min_value, type_min), std::min(a.max_value, type_max)};
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/arith/const_int_bound.cc:372]: Fixed in `a8349e3` — Cast bounds now distinguish signed (`[-2^(n-1), 2^(n-1)-1]`) from unsigned (`[0, 2^n-1]`) integer types using `IsUnsignedInt()`. Added tests for both cases.

File: src/ir/arith/const_int_bound.cc [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+    if (!exp.is_const() || exp.min_value < 0) return Everything();
+    int64_t e = exp.min_value;
+    if (e == 0) return {1, 1};
+    if (e == 1) return base;
+    if (base.is_non_negative()) {
+      // Monotonically increasing for non-negative base
+      int64_t lo = 1, hi = 1;
+      for (int64_t i = 0; i < e; ++i) {
+        lo = InfAwareMul(lo, base.min_value);
+        hi = InfAwareMul(hi, base.max_value);
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/arith/const_int_bound.cc]: Fixed in `a8349e3` — same as gemini comment: replaced with O(log e) exponentiation by squaring.

File: tests/ut/ir/arith/test_const_int_bound.py [hunk 6]
```diff
@@ -0,0 +1,506 @@
...
+from pypto import DataType, ir
+from pypto.arith import ConstIntBound, ConstIntBoundAnalyzer
+
+S = ir.Span.unknown()
+INT = DataType.INT64
+
+
+def make_var(name: str) -> ir.Var:
+    return ir.Var(name, ir.ScalarType(INT), S)
+
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply tests/ut/ir/arith/test_const_int_bound.py:22]: Fixed in `a8349e3` — added 5 edge case tests: `test_neg_int64_min`, `test_abs_int64_min`, `test_large_pow_exponent`, `test_cast_to_unsigned`, `test_cast_to_signed`.

File: python/bindings/modules/arith.cpp [hunk 7]
```diff
@@ -61,6 +65,40 @@ void BindArith(nb::module_& m) {
...
       },
       nb::arg("a"), nb::arg("b"), "Extended Euclidean: returns (gcd, x, y) where a*x + b*y = gcd.");
+
+  // ConstIntBound
+  nb::class_<ir::arith::ConstIntBound>(arith, "ConstIntBound",
+                                       "Inclusive integer bounds [min_value, max_value] for an expression.")
+      .def_ro("min_value", &ir::arith::ConstIntBound::min_value, "Inclusive lower bound.")
+      .def_ro("max_value", &ir::arith::ConstIntBound::max_value, "Inclusive upper bound.")
+      .def_ro_static("kPosInf", &ir::arith::ConstIntBound::kPosInf, "Sentinel for positive infinity.")
+      .def_ro_static("kNegInf", &ir::arith::ConstIntBound::kNegInf, "Sentinel for negative infinity.")
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply python/bindings/modules/arith.cpp:77]: Fixed in `e5b4a0b` — added `nb::init<int64_t, int64_t>()` constructor to `ConstIntBound` binding and updated the type stub.

File: src/ir/arith/const_int_bound.cc [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+
+ protected:
+  // --- Leaf nodes ---
+
+  Bound VisitExpr_(const ConstIntPtr& op) override { return {op->value_, op->value_}; }
+
+  Bound VisitExpr_(const ConstFloatPtr& op) override {
+    if (std::isnan(op->value_) || std::isinf(op->value_)) return Everything();
+    return {static_cast<int64_t>(std::floor(op->value_)), static_cast<int64_t>(std::ceil(op->value_))};
+  }
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/arith/const_int_bound.cc:194]: Fixed in `e5b4a0b` — added range check before `static_cast<int64_t>`: finite doubles outside int64 range now clamp to `kNegInf`/`kPosInf` instead of triggering UB.

File: src/ir/arith/const_int_bound.cc [hunk 9]
```diff
@@ -0,0 +1,473 @@
...
+std::function<void()> ConstIntBoundAnalyzer::Impl::EnterConstraint(const ExprPtr& constraint) {
+  std::vector<std::pair<const Expr*, Bound>> recovery;
+
+  // Helper: try to tighten bound for a variable.
+  auto TryTighten = [&](const Expr* var_ptr, const Bound& new_bound) {
+    auto it = var_map_.find(var_ptr);
+    Bound old = (it != var_map_.end()) ? it->second : Everything();
+    recovery.emplace_back(var_ptr, old);
+    var_map_[var_ptr] = {std::max(old.min_value, new_bound.min_value),
+                         std::min(old.max_value, new_bound.max_value)};
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/arith/const_int_bound.cc:393]: Fixed in `e5b4a0b` — recovery now iterates in reverse (`rbegin`/`rend`) so compound constraints like `x >= 0 && x < 8` correctly restore the original pre-scope bound.
